Make hyperv powershell commands work in non-interactive ssh session

pdostal opened the pull request:
As some Hyper-V PowerShell commands in Windows Server 2019 require
interactive SSH session we need to change this behavior using
special variables provided in each SSH call.

- os-autoinst: os-autoinst/os-autoinst-distri-opensuse#12098
- Verification run: [Win2k19](http://pdostal-server.suse.cz/tests/11198) [Win2k16](http://pdostal-server.suse.cz/tests/11212) [Win2k12r2](http://pdostal-server.suse.cz/tests/11211)

Martchus reviewed (approved):
Looks good. I'm only adding the "acceptance-tests-needed" tag because currently one of the verification runs is still pending.

okurz reviewed (changes requested):
http://pdostal-server.suse.cz/tests/11204/file/autoinst-log.txt turned out to be incomplete:

```
[0m[32m[2021-03-09T15:49:12.700 CET] [debug] >>> testapi::_check_backend_response: match=text-logged-in-root,tty5-selected timed out after 180 (assert_screen)
[0m[37m[2021-03-09T15:49:12.729 CET] [debug] post_fail_hook failed: no candidate needle with tag(s) 'tty5-selected, text-logged-in-root' matched
[0m[37m[2021-03-09T15:49:12.731 CET] [debug] Loading a VM snapshot lastgood
[0m[2021-03-09T15:49:12.732 CET] [debug] <<< backend::baseclass::run_ssh_cmd(cmd="powershell -Command \"\$ProgressPreference='SilentlyContinue'; Restore-VMSnapshot -VMName openQA-SUT-8 -Name lastgood -Confirm:\$false\"", keep_open=1, wantarray=0)
[2021-03-09T15:49:12.732 CET] [debug] <<< backend::baseclass::run_ssh(cmd="powershell -Command \"\$ProgressPreference='SilentlyContinue'; Restore-VMSnapshot -VMName openQA-SUT-8 -Name lastgood -Confirm:\$false\"", keep_open=1, wantarray=0)
[2021-03-09T15:49:12.732 CET] [debug] <<< backend::baseclass::new_ssh_connection(keep_open=1, blocking=1, wantarray=0)
[33m[2021-03-09T16:05:26.490 CET] [info] ::: backend::baseclass::die_handler: Backend process died, backend errors are reported below in the following lines:
Unable to create SSH channel for executing "powershell -Command "$ProgressPreference='SilentlyContinue'; Restore-VMSnapshot -VMName openQA-SUT-8 -Name lastgood -Confirm:$false"": no libssh2 error registered at /usr/lib/os-autoinst/backend/baseclass.pm line 1416.

[0m[37m[2021-03-09T16:05:26.490 CET] [debug] Closing SSH serial connection with openqaw5-xen-1.qa.suse.de
[0m[37m[2021-03-09T16:05:26.491 CET] [debug] Destroying openQA-SUT-8 virtual machine
[0m[2021-03-09T16:05:26.491 CET] [debug] <<< backend::baseclass::run_ssh_cmd(cmd="powershell -Command Stop-VM -Force -VMName openQA-SUT-8 -TurnOff", keep_open=1, wantarray=0)
[2021-03-09T16:05:26.491 CET] [debug] <<< backend::baseclass::run_ssh(cmd="powershell -Command Stop-VM -Force -VMName openQA-SUT-8 -TurnOff", wantarray=0, keep_open=1)
[2021-03-09T16:05:26.491 CET] [debug] <<< backend::baseclass::new_ssh_connection(blocking=1, wantarray=0, keep_open=1)
[31m[2021-03-09T16:05:26.492 CET] [warn] !!! backend::baseclass::run_capture_loop: capture loop failed Unable to create SSH channel for executing "powershell -Command Stop-VM -Force -VMName openQA-SUT-8 -TurnOff": Unable to send channel-open request (-7 LIBSSH2_ERROR_SOCKET_SEND) at /usr/lib/os-autoinst/backend/baseclass.pm line 1416.

[0m[37m[2021-03-09T16:05:26.492 CET] [debug] sending magic and exit
[0m[37m[2021-03-09T16:05:26.492 CET] [debug] received magic close
```

okurz commented on t/22-svirt.t:
```diff
@@ -193,6 +193,9 @@ subtest 'SSH usage in svirt' => sub {
     is_deeply(\@output, [0, 'foo', ''], 'Command successful exit with output');
+    # Check Hyper-V command
+    my $vmname = 'Test-VM-1';
+    is($svirt->run_ssh_cmd(qq($ps "\$ProgressPreference='SilentlyContinue'; Remove-VM -Force -VMName $vmname"), 0, 'Hyper-V command successful exit');
```
from test results:

```
[  136s] 3: Global symbol "$ps" requires explicit package name (did you forget to declare "my $ps"?) at ./22-svirt.t line 198.
[  136s] 3: Type of arg 1 to Test::Exception::dies_ok must be block or sub {} (not reference constructor) at ./22-svirt.t line 286, near ""die() when log file wasn't created")"
[  136s] 3: Type of arg 1 to Test::Exception::dies_ok must be block or sub {} (not reference constructor) at ./22-svirt.t line 291, near "'die() when emulate CONSOLE_EXIT token in log file')"
[  136s] 3: Execution of ./22-svirt.t aborted due to compilation errors.
[  136s] 3: [07:44:48] ./22-svirt.t ............................. 
```

pdostal commented:
> http://pdostal-server.suse.cz/tests/11204/file/autoinst-log.txt turned out to be incomplete:

The worker was in my opinion overtaken by another test from another instance. It succeeded after restart.

> Looks good. I'm only adding the "acceptance-tests-needed" tag because currently one of the verification runs is still pending.

I added one more test case but honestly I don't know what else to add.

I also added a VR on Win2k12r2.

okurz reviewed (approved):
Verification runs are all fine so we can regard this as acceptable. Thank you so far. We would really appreciate if you could also help increase the test coverage for the files as reported by codecov. E.g. with mocked methods have a unit test statement coverage for the files you changed. This can be done in a follow-up PR.

Martchus commented:
Your passed verification run has the reason:

>  Result: passed finished about 8 hours ago ( 01:26 hours )
Reason: backend done: Unable to create SSH channel for executing "powershell -Command "if ($(Get-VM -VMName openQA-SUT-8 | Where-Object {$_.state -eq 'Off'})) { exit 1 } else { exit 0 }"": no libssh2 error registered at /usr/lib/os-autoinst/backend/baseclass.pm line 1416. 

That's a little bit strange but likely the pool directory just wasn't clean when running the job.

okurz reviewed (changes requested):
oh, good catch. I would prefer if this is being looked into first anyway. At best the case of unclean pool dir can be handled as well.

mloviska commented:
> Your passed verification run has the reason:
> 
> > Result: passed finished about 8 hours ago ( 01:26 hours )
> > Reason: backend done: Unable to create SSH channel for executing "powershell -Command "if ($(Get-VM -VMName openQA-SUT-8 | Where-Object {$_.state -eq 'Off'})) { exit 1 } else { exit 0 }"": no libssh2 error registered at /usr/lib/os-autoinst/backend/baseclass.pm line 1416.
> 
> That's a little bit strange but likely the pool directory just wasn't clean when running the job.

Look like the same issue as we can see in vmware tests described in https://progress.opensuse.org/issues/81878 or in https://progress.opensuse.org/issues/44771

Martchus reviewed (changes requested):
The auto-generated Makefiles should not be checked in. Note that the project moved to CMake now so we've cleaned up `.gitignore`.

pdostal commented:
@okurz @Martchus this PR should now be ready. I will try to improve the test coverage in another PR.

Martchus reviewed (approved):
Ok. About extending tests: There's already `22-svirt.t` which might serve as a starting point.

okurz commented:
@pdostal can you comment on https://github.com/os-autoinst/os-autoinst/pull/1631#issuecomment-795153394 regardin the errors encountered in the verification runs even though they don't fail?

pdostal commented:
> @pdostal can you comment on [#1631 (comment)](https://github.com/os-autoinst/os-autoinst/pull/1631#issuecomment-795153394) regardin the errors encountered in the verification runs even though they don't fail?

I don't know much more than @mloviska already stated. This issue is not related to this PR, is not Hyper-V specific and is handled in [poo#81878](https://progress.opensuse.org/issues/81878) separately.

Martchus commented:
And the fact that the *passed* test shows that reason is likely just due to a dirty pool directory or some other unrelated setup problems.

okurz reviewed (approved):
ok then :)
